internal/webhooks/clusterclass.go

```diff
@@ -154,20 +163,210 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable schema change for %s is invalid. It failed validation in Clusters %v", variable, clusters),
+				))
+		}
+	}
+	if len(errs.added) > 0 {
+		for variable, clusters := range errs.added {
+			allErrs = append(allErrs,
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable %s can not be added. It failed validation in Clusters %v", variable, clusters),
+				))
+		}
+	}
+	if len(errs.required) > 0 {
+		for variable, clusters := range errs.required {
+			allErrs = append(allErrs,
+				field.Forbidden(path.Index(clusterClassVariablesIndex[variable]),
+					fmt.Sprintf("variable %v is required but is not defined in Clusters %v", variable, clusters),
+				))
+		}
+	}
+	return allErrs
+}
+
+func clustersWithoutVar(allClusters, clustersWithVar []string) []string {
+	clustersWithVarMap := make(map[string]interface{})
+	for _, cluster := range clustersWithVar {
+		clustersWithVarMap[cluster] = nil
+	}
+	var clusterDiff []string
+	for _, cluster := range allClusters {
+		if _, found := clustersWithVarMap[cluster]; !found {
+			clusterDiff = append(clusterDiff, cluster)
+		}
+	}
+	return clusterDiff
+}
+
+// variableClusterReferences contains variable names and an associated list of clusters which reference them.
+type variableClusterReferences map[string][]string
+
+// add appends a value to the given array and creates the array if it does not already exist.
+func (v variableClusterReferences) add(varName string, clusterName ...string) {
+	v[varName] = append(v[varName], clusterName...)
+}
+
+func (v variableClusterReferences) get(varName string) []string {
+	return v[varName]
+}
+
+// clusterClassVariablesForValidation used to define changes between two ClusterClasses.
+type clusterClassVariablesForValidation struct {
+	altered  map[string]*clusterv1.ClusterClassVariable
+	added    map[string]*clusterv1.ClusterClassVariable
+	removed  map[string]*clusterv1.ClusterClassVariable
+	required map[string]*clusterv1.ClusterClassVariable
+}
+
+// getClusterClassVariablesForValidation returns a struct with the four classes of ClusterClass Variables that require validation:
+// - added which have been added to the ClusterClass on update.
+// - altered which have had their schemas changed on update.
+// - removed which have been removed from the ClusterClass on update.
+// - required (with 'required' : true) from the new ClusterClass.
+func getClusterClassVariablesForValidation(oldVars, newVars map[string]*clusterv1.ClusterClassVariable) clusterClassVariablesForValidation {
+	// Create a map each for new, altered,  removed and required Variables.
+	addedVariables := map[string]*clusterv1.ClusterClassVariable{}
+	alteredVariables := map[string]*clusterv1.ClusterClassVariable{}
+	removedVariables := map[string]*clusterv1.ClusterClassVariable{}
+	requiredVariables := map[string]*clusterv1.ClusterClassVariable{}
```

Comment:
Same as above, consider having the action type of the variable defined somewhere else rather than carrying around multiple maps.

We should try to avoid having "business logic"/semantic be part of the variable name, but rather have the semantic behavior as an input

Reply:
I've added a variableValidationKey as the map key here so this is now a single map where the key has the action type + variable name. This simplifies the logic a bit and gets rid of keeping the logic in the variable name - let me know what you think.

Reply:
To me the change has the following downsides:
* `added`, `altered`, `removed` do not sound like `variableValidationAction`'s to me. `variableValidationAction` sounds to me like it's the type of the `validation`, which does only fit for `required`. Or maybe it is actually the cause why we want to validate something, but then `variableValidationAction` sound strange to me.
* Previously the errors where ordered, i.e. all removed, all altered, all added all required errors. I think now it's random and thus also non-deterministic.
* The new code is harder to read and understand.

But maybe that's just me. If you'll get consensus on the current implementation, I don't want to block.

Reply:
Agreed that there are definite downsides to this approach. 
- The naming / additional struct aren't ideal - but this is the 'type' of vallidation being done to my mind: We have four validation paths and each of these keys leads to some validation action being performed based on the 'type'. To me there's no way to really separate the cause/action categorically here, but that leads to some confusion - have you another idea on naming? 

- I don't think the ordering is important due to the way errors are being aggregated though - we're still reporting the index in the error message, but I don't think we should care about the order we're printing those messages.

- Agreed this is probably harder to read, but gets rid of the issue with using specific map names to define actions. I'm happy to go either direction on the structure here.

@vincepri WDYT?

Reply:
> have you another idea on naming?

Best idea I can come up with is something like `validationType`, but that's just potentially better because it's more generic and thus less expressive / potentially less wrong :)

> I don't think the ordering is important due to the way errors are being aggregated though - we're still reporting the index in the error message, but I don't think we should care about the order we're printing those messages.

I just thought as a user faced with a wall of errors it might drive me crazy to get the different error types in mixed order and in a different order each time I run against the webhook (aka when I'm working my way through fixing those errors). Imagine golangci-lint gives you a long list of errors, each time in a different order and the types are mixed (maybe that's the current behavior in fact :P).

Reply:
I hadn't thought of it that way - I've made an update to make the error order defined (in terms of error type, variable impacted is still unordered inside that category). I've also update the name of the to variableValidationType to be more generic.

Reply:
Looks good from my side.